Fix clickable legend filtering with multiple metrics

w33ble opened the pull request:
Fixes #4793 
- Attach aggConfig to the values
- Use aggConfig to find the matching labels

![2015-09-14 14_14_45](https://cloud.githubusercontent.com/assets/404731/9862028/fba6a748-5aea-11e5-8a6f-c32de18e25a6.gif)

epixa commented on src/ui/public/filter_bar/filter_bar_click_handler.js:
```diff
@@ -17,32 +25,21 @@ define(function (require) {
-          if (isLegendLabel) results = _.filter(results, findLabel); // filter results array by legend label
+          // For legend clicks, use the last bucket in the path
+          if (isLegendLabel) {
+            aggBuckets = _.slice(aggBuckets, aggBuckets.length - 1);
```
How about `_.last(appBuckets)`?

epixa replied:
Ah, I see... you actually want an array here.

epixa replied:
I'm generally extremely wary about extracting values from an array based on the order of the array, but I don't see a better alternative in this case.  It looks like the tests for `AggConfigResult#getPath()` do specifically verify order, so we're probably alright.

w33ble replied:
Yeah, the order is really important in getPath - if the order breaks, it'll affect a lot more than this.

epixa commented:
LGTM

Bargs commented:
I'm noticing some odd behavior:

* Clicking the legend has a different effect than clicking its corresponding bar in the chart. In my example below the legend is only filtering by my "term" bucket but not the date. Clicking the bar filters by both term and date.

* Filtering by a term uncovers an instance of that term that wasn't previously displayed in the graph. This is probably unrelated to the clickable legend but it's really odd.

* When reproducing the bug, the clickable legends didn't seem to work with just a single metric.

![filterlegend](https://cloud.githubusercontent.com/assets/6239176/9939710/bec4f6d0-5d39-11e5-8fa0-af8a11e6d795.gif)

Bargs commented:
Scratch all that, I just realized the second bucket was limited to 5 results, explaining why the extra instance of that term wasn't showing up until I filtered on it. That also explains why the clickable legend works a little differently than clicking on an individual bar. Might still be a little confusing to a user... but at least it's logically consistent. 

This PR LGTM now.

Bargs commented:
Although, why does clicking a single legend icon incrementally reduce the time range?

![filterlegend](https://cloud.githubusercontent.com/assets/6239176/9941080/115cc2a4-5d41-11e5-8ca4-2f5af14752b8.gif)

w33ble commented:
> why does clicking a single legend icon incrementally reduce the time range?

Because the last (and only) bucket in the agg is a date histogram, which we treat internally differently than any other bucket. Instead of creating a time filter, we just adjust the time series. This happens when you click on the graph as well.

Bargs commented:
The graph makes sense since you're selecting a specific slice of time, but
the legend is odd. I would expect the legend to be unclickable in this
situation.

On Thu, Sep 17, 2015, 2:04 PM Joe Fleming <notifications@github.com> wrote:

> why does clicking a single legend icon incrementally reduce the time range?
>
> Because the last (and only) bucket in the agg is a date histogram, which
> we treat internally differently than any other bucket. Instead of creating
> a time filter, we just adjust the time series. This happens when you click
> on the graph as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/kibana/pull/4940#issuecomment-141168562>.
>

w33ble commented:
Ah, yeah, I think you're right. We're only disabling clicking the legend when the label is explicitly "Count".

I need to change this fix anyway, so I'll see if I can address that as well.

w33ble commented:
I changed the implementation so that it work regardless of the aggConfig order. This also fixed the time range change when the metric is not simply *Count*

![2015-09-18 15_07_51](https://cloud.githubusercontent.com/assets/404731/9972201/11b41258-5e17-11e5-9aeb-bcbd01197462.gif)
![2015-09-18 15_00_51](https://cloud.githubusercontent.com/assets/404731/9972202/11c31636-5e17-11e5-97d2-bdaaa6ffa8e4.gif)

Bargs commented:
Non "count" metrics are still clickable but it throws an error which prevents other legend items from working. The only way to get out of this broken state seems to be to reload the visualization.

![filterlegend](https://cloud.githubusercontent.com/assets/6239176/9996291/55014b0e-6054-11e5-9c1a-8d0d247a37e8.gif)

This is the error I'm getting in the console:

```
TypeError: Cannot read property 'aggConfig' of undefined
    at findByParam (filter_bar_click_handler.js:17)
    at Function.module.exports (filter_bar_click_handler.js:49)
    at events.js:45
    at deferred.promise.then.wrappedCallback (angular.js:11682)
    at angular.js:11768
    at Scope.promises.$get.Scope.$eval (angular.js:12811)
    at Scope.promises.$get.Scope.$digest (angular.js:12623)
    at angular.js:12849
    at completeOutstandingRequest (angular.js:4411)
    at angular.js:4725
```

Bargs commented:
Looks pretty good to me. The only thing I noticed is that the non-Count metrics still get a pointer when you hover over them in the legend and their click event still gets handled, unlike Count. Looks like this is because Count is given special treatment in legend.js:

https://github.com/elastic/kibana/blob/master/src/ui/public/vislib/lib/legend.js#L170
https://github.com/elastic/kibana/blob/master/src/ui/public/vislib/lib/legend.js#L211

I don't think it's a showstopper, but testing for "Count" like this will make things more fragile. If it's out of scope for this ticket we might want to create a new one for making legend.js more robust.

Assigning to Spencer for his opinion and further review.

spalger commented:
@Bargs since this pull properly fixes the stated issue I think we should create a new issue to track the css-cursor bug.

spalger commented:
Opened https://github.com/elastic/kibana/issues/4992 to track ^
